### ISSUE TITLE
[Question] Concurrent writes to the PipeWriter

### ISSUE THREAD

**Issue by Cogitri** [arcnmx/pipe-rs#8]
Hello,

first off, thanks a lot for the crate, it really helped me a lot! :)

I wanted to know if it's possible to do concurrent writes to  a `PipeWriter` - right now I wrap it in a `Mutex` to make sure I don't write to it from two different threads at the same time to not mix two unrelated messages, but is that necessary?

Thanks!

**Comment by arcnmx** [comment]
`PipeWriter` can be `.clone()`d, so you could just have each thread own its own handle; there's no need for additional synchronization. Of course you'll need to be aware of the multi-threading caveats in that you need to make sure your writes are atomic (only call `write()` with an entire valid packet, and be careful of how you buffer writes) or the `PipeReader` might get confused interleaved data. You may want to look into just using `crossbeam_channel` instead if you're doing complex concurrency stuff - threaded byte streams are generally just going to be an unnecessary mess.

Basically, you can avoid the `Mutex`, but should make sure all calls to `pipeWriter.write()` contain complete messages. You can write them into a vec first if you don't have full control over the writer code.

A few non-&mut fns/impls have been added to the writer to make this more straightforward to do.

(sorry for the late response)

**Comment by Cogitri** [comment]
Thanks for the answer :)
